cpp_exercise/src/lib/ipcshmlib.cc

```diff
@@ -0,0 +1,247 @@
+#include"ipcshmlib.h"
+#include "ipcexceptionlib.h"
+
+
+/**
+ * @brief Shared Memory Class
+ * 
+ */
+
+
+/**
+ * Construct a new Shared Memory Sender:: Shared Memory Sender object
+ * 
+ * 
+ */
+SharedMemorySender::SharedMemorySender(const std::string filename): _file_name(filename){
+        std::cout<<"**************************************************************"<<std::endl;
+        std::cout<<"****************************IPCSender: SharedMemory PROTOCOL**************************"<<std::endl;
+        std::cout<<"**************************************************************"<<std::endl;
+        std::cout<<"**************************************************************"<<std::endl;
+
+
+    
+
+    while(_shm_fd == -1){
+     std::cout<<"          Waiting for the Receiver... "<<std::endl;
+     _shm_fd = shm_open(_shm_name.c_str(), O_RDWR, 0666);
+     _sem_sender = sem_open(_semaphoresender_name.c_str(),0);
+     _sem_receiver = sem_open(_semaphorereceiver_name.c_str(),0);
+
+         sleep(5);
+    }
+    
+    _err = ftruncate(_shm_fd, _shm_size);
+    if(_err<0){
+         std::cout<<strerror(errno)<<std::endl;
+         throw IPCException("IPCSender ERROR: Cannot truncate the Shared Memory.");
+
+    }
+ 
+   
+
+    _ptr = (struct _shm_data_struct *)mmap(0, _shm_size, PROT_WRITE, MAP_SHARED, _shm_fd, 0);
+
+    if(_ptr == (struct _shm_data_struct *)-1){
+        std::cout<<strerror(errno)<<std::endl;
+        throw IPCException("IPCSender ERROR: Cannot map the Shared Memory.");
+    }
+    
+    
+    FileHandler fd(_file_name);
+
+    fd.openForReading();
+
+    std::cout<<"          File Name: "<<_file_name<<std::endl;
+    std::cout<<"          File Size: "<<fd.getSize()<<" bytes"<<std::endl;
+
+    
+    std::cout<<"          SharedMemory is Open: Trying to send the file  "<<std::endl;
+  
+
+    while(true){
+
+                
+       _err = sem_wait(_sem_receiver);
+       if(_err<0){
+         std::cout<<strerror(errno)<<std::endl;
+         throw IPCException("IPCSender ERROR: Semaphore receiver.");
+
+         }   
+
+        std::vector<unsigned char>  buffer; 
+        
+        buffer = fd.readFile();
+
+        
+        _shm_data_struct tempstruct;
+        tempstruct.datasize = buffer.size();
+
+        std::copy(buffer.begin(), buffer.end(),tempstruct.data);
```

Comment:
you don't need this intermediate copy, it's possible to copy directly from buffer to ptr

Reply:
since I transfer the messages in struct the copy function copy the data of buffer in the an array member of the struct. 
if I wanna do it directly I should use pointer instead of array which cause other kinds of errors and bugs later